src/braket/circuits/circuit.py

```diff
@@ -773,6 +805,91 @@ def apply_initialization_noise(
 
         return apply_noise_to_moments(self, noise, target_qubits, "initialization")
 
+    def make_bound_circuit(self, param_values: Dict[str, Number], strict: bool = False) -> Circuit:
+        """
+        Binds FreeParameters based upon their name and values passed in. If parameters
+        share the same name, all the parameters of that name will be set to the mapped value.
+
+        Args:
+            param_values (Dict[str, Number]):  A mapping of FreeParameter names
+                to a value to assign to them.
+            strict (bool, optional): If True, raises a ValueError if none of the FreeParameters
+                in param_values appear in the circuit. False by default."
+
+        Returns:
+            Circuit: Returns a circuit with all present parameters fixed to their respective
+            values.
+        """
+        if strict:
+            self._validate_parameters(param_values)
+        return self._use_parameter_value(param_values)
+
+    def _validate_parameters(self, parameter_values: Dict[str, Number]):
+        """
+        This runs a check to see that the parameters are in the Circuit.
+
+        Args:
+            parameter_values (Dict[str, Number]):  A mapping of FreeParameter names
+                to a value to assign to them.
+
+        Raises:
+            ValueError: If there are no parameters that match the key for the arg
+            param_values.
+        """
+        parameter_strings = set()
+        for parameter in self.parameters:
+            parameter_strings.add(str(parameter))
+        for param in parameter_values:
+            if param not in parameter_strings:
+                raise ValueError(f"No parameter in the circuit named: {param}")
+
+    def _use_parameter_value(self, param_values: Dict[str, Number]) -> Circuit:
+        """
+        Creates a Circuit that uses the parameter values passed in.
+
+        Args:
+            param_values (Dict[str, Number]): A mapping of FreeParameter names
+                to a value to assign to them.
+
+        Returns:
+            Circuit: A Circuit with specified parameters swapped for their
+            values.
+
+        """
+        fixed_circ = Circuit()
+        for instruction in self.instructions:
+            if self._check_for_params(instruction):
+                params = instruction.operator.parameters
+                values = list(
+                    map(lambda x: param_values[str(x)] if str(x) in param_values else x, params)
+                )
+                for val in values:
+                    if not isinstance(val, FreeParameter):
+                        self._validate_parameter_value(val)
+                fixed_circ.add(
+                    Instruction(type(instruction.operator)(*values), target=instruction.target)
+                )
```

Comment:
I'm a bit iffy that

```python
Instruction(type(instruction.operator)(*values), target=instruction.target)
```

relies on the order of `instruction.operator.parameters` matching that of the arguments of `instruction.operator`.

Thinking on it a bit more, I'm thinking that base parameterizable classes, like `AngledGate` and [`PauliNoise`](https://github.com/aws/amazon-braket-sdk-python/blob/main/src/braket/circuits/noise.py#L216), should have specific methods to bind parameters and create new, bound instances. With such a method, let's call it `bind_values`, this would look something like

```python
fixed_circ.add(Instruction(instruction.operator.bind_values(...)), instruction.target)
```

Reply:
I like the idea that we want an explicit method, but it should probably create a new Instruction rather than mutating the existing one.

Reply:
Yes, this would return a new object of the type with the new values kept in. We need the existing object to remain not mutated in the circuit so that later on it can be reused.

Reply:
Yep, by fixing parameters, I'm assuming a new object is created

Reply:
I'm usually thinking functionally, so I'll explicitly mention mutation. In this case, it becomes a nice extension of [currying](https://en.wikipedia.org/wiki/Currying)

Reply:
Now that `bind_values` exists, this entire block can be simplified:
```suggestion
                fixed_circ.add(
                    Instruction(
                        instruction.operator.bind_value(*param_values), target=instruction.target
                    )
                )
```
or something along these lines.

Reply:
fixed